controllers/storageclaim_controller.go

```diff
@@ -501,64 +543,103 @@ func (r *StorageClaimReconciler) getCephRBDStorageClass(data map[string]string)
 	return storageClass
 }
 
-func (r *StorageClaimReconciler) getCephFSVolumeSnapshotClass(data map[string]string) *snapapi.VolumeSnapshotClass {
+func (r *StorageClaimReconciler) getCephDriverVolumeSnapshotClass(driverType string,
+	data map[string]string) *snapapi.VolumeSnapshotClass {
+	var driverName string
+	switch driverType {
+	case "cephfs":
+		driverName = csi.GetCephFSDriverName()
+	case "rbd":
+		driverName = csi.GetRBDDriverName()
+	}
 	volumesnapshotclass := &snapapi.VolumeSnapshotClass{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: r.storageClaim.Name,
 		},
-		Driver:         csi.GetCephFSDriverName(),
+		Driver:         driverName,
 		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
 		Parameters:     data,
 	}
 	return volumesnapshotclass
 }
 
-func (r *StorageClaimReconciler) getCephRBDVolumeSnapshotClass(data map[string]string) *snapapi.VolumeSnapshotClass {
-	volumesnapshotclass := &snapapi.VolumeSnapshotClass{
+func (r *StorageClaimReconciler) getCephDriverVolumeGroupSnapshotClass(driverType string,
+	data map[string]string) *groupsnapapi.VolumeGroupSnapshotClass {
+	var driverName string
+	switch driverType {
+	case "cephfs":
+		driverName = csi.GetCephFSDriverName()
+	case "rbd":
+		driverName = csi.GetRBDDriverName()
+	}
+	volumegroupsnapshotclass := &groupsnapapi.VolumeGroupSnapshotClass{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: r.storageClaim.Name,
 		},
-		Driver:         csi.GetRBDDriverName(),
+		Driver:         driverName,
 		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
 		Parameters:     data,
 	}
-	return volumesnapshotclass
+	return volumegroupsnapshotclass
 }
 
-func (r *StorageClaimReconciler) createOrReplaceStorageClass(storageClass *storagev1.StorageClass) error {
-	existing := &storagev1.StorageClass{}
-	existing.Name = r.storageClaim.Name
+func (r *StorageClaimReconciler) createOrReplaceClass(class client.Object) error {
```

Comment:
Shouldn't we get the existing before the switch case? existingParameters will always be empty in this case. 
We should, 
1. set the name, to existing, get it based on the object kind
2. compare the existing parameter/labels/annotations to the desired
3. if equal break
4. own the desired
5. if UID is not nil delete it
6. create the desired

Could we also rename class client.Object to a better name? maybe to desired?